hudi-spark-datasource/hudi-spark3.3.x/src/main/scala/org/apache/spark/sql/execution/datasources/HoodieSpark33PartitionedFileUtils.scala

```diff
@@ -24,12 +24,15 @@ import org.apache.hudi.storage.StoragePath
 import org.apache.hadoop.fs.FileStatus
 import org.apache.spark.sql.catalyst.InternalRow
 
+import java.net.URI
+
 /**
  * Utils on Spark [[PartitionedFile]] and [[PartitionDirectory]] for Spark 3.3.
  */
 object HoodieSpark33PartitionedFileUtils extends HoodieSparkPartitionedFileUtils {
   override def getPathFromPartitionedFile(partitionedFile: PartitionedFile): StoragePath = {
-    new StoragePath(partitionedFile.filePath)
+    //see SPARK-41970
+    new StoragePath(new URI(partitionedFile.filePath))
```

Comment:
So Spark 3.4 and 3.5 do not have the problem?

Reply:
https://issues.apache.org/jira/browse/SPARK-41970 Nope, spark changed partitioned file in spark 3.4.
From the issue ticket:
```
Today, Spark represents file paths in various ways. Sometimes they are Hadoop `Path`s, sometimes they are `Path.toString`s, and sometimes they are `Path.toUri.toString`s.

This discrepancy means that Spark does not always work when user provided strings have special characters. Sometimes Spark will try to create a URI with an unescaped string; sometimes Spark will double-escape a path and try to access the wrong file.

 

This issue proposes a new `SparkPath` class meant to provide type safety when Spark is dealing with paths.
```
This is exactly the problem that was happening here. The encoding and decoding wasn't consistent.

Reply:
Got it